OpenRA.Game/Graphics/ShaderBindings.cs

```diff
@@ -0,0 +1,57 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.IO;
+
+namespace OpenRA.Graphics
+{
+	public readonly struct ShaderVertexAttribute
+	{
+		public readonly string Name;
+		public readonly int Index;
+		public readonly int Components;
+		public readonly int Offset;
+
+		public ShaderVertexAttribute(string name, int index, int components, int offset)
+		{
+			Name = name;
+			Index = index;
+			Components = components;
+			Offset = offset;
+		}
+	}
+
+	public abstract class ShaderBindings : IShaderBindings
+	{
+		public string VertexShaderName { get; }
+		public string VertexShaderCode { get; }
+		public string FragmentShaderName { get; }
+		public string FragmentShaderCode { get; }
+		public int Stride { get; }
+
+		public abstract ShaderVertexAttribute[] Attributes { get; }
+
+		protected ShaderBindings(string name)
+		{
+			Stride = Attributes[^1].Offset + Attributes[^1].Components * 4;
```

Comment:
Perhaps we should add some validation to ensure attributes are in order, without overlaps, etc.